When deserializing using `ObjectMapper::treeToValue`, NPE occurs when `JsonParser::readValueAs` is called

### Search before asking

- [x] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

The issue occurs as described in the title.

This error does not occur when using `readValue` or when the JSON representation is numeric.

It appears there are differences in the processing path for `treeToValue` compared to `readValue`, resulting in `DeserializationContext._readCapabilities` being `null`.
Furthermore, access to this property only occurs when conversion from a string to a primitive is performed.
Specifically, it is called from the `StdDeserializer._checkFromStringCoercion` method.

### Version Information

3.0.0 ~ 3.1.0-RC1(9720c26c3cac2578a35f9858f497333a0a78dc20)

### Reproduction

`OK` will be printed, but processing `NG` will cause an exception.

```java
package tools.jackson.databind;

import com.fasterxml.jackson.annotation.JsonValue;
import org.junit.jupiter.api.Test;
import tools.jackson.core.JacksonException;
import tools.jackson.core.JsonParser;
import tools.jackson.databind.annotation.JsonDeserialize;
import tools.jackson.databind.deser.std.StdDeserializer;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class Kotlin1089Test {
    @JsonDeserialize(using = Inner.Deser.class)
    public record Inner(@JsonValue long value) {
        static class Deser extends StdDeserializer<Inner> {
            protected Deser() { super(Inner.class); }

            @Override
            public Inner deserialize(JsonParser p, DeserializationContext ctxt) throws JacksonException {
                return new Inner(p.readValueAs(Long.class));
            }
        }
    }

    public record Outer(Inner inner) { }

    @Test
    public void testSerialization() throws Exception {
        var mapper = new ObjectMapper();
        var json = "{\"inner\":\"123\"}";

        var ok = mapper.readValue(json, Outer.class);
        System.out.println("OK: " + ok);

        var tree = mapper.readTree(json);
        var ng = mapper.treeToValue(tree, Outer.class);
        System.out.println("NG: " + ng);
    }
}
```

The stack trace is as follows.

```
tools.jackson.databind.DatabindException: Cannot invoke "tools.jackson.core.util.JacksonFeatureSet.isEnabled(tools.jackson.core.util.JacksonFeature)" because "this._readCapabilities" is null
 at [No location information] (through reference chain: tools.jackson.databind.Kotlin1089Test$Outer["inner"])

	at tools.jackson.databind/tools.jackson.databind.DatabindException.wrapWithPath(DatabindException.java:111)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1936)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:860)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:721)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1503)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:598)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializer.deserialize(BeanDeserializer.java:200)
	at tools.jackson.databind/tools.jackson.databind.deser.DeserializationContextExt.readRootValue(DeserializationContextExt.java:266)
	at tools.jackson.databind/tools.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:2610)
	at tools.jackson.databind/tools.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:978)
	at tools.jackson.databind/tools.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:1291)
	at tools.jackson.databind/tools.jackson.databind.Kotlin1089Test.testSerialization(Kotlin1089Test.java:36)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.NullPointerException: Cannot invoke "tools.jackson.core.util.JacksonFeatureSet.isEnabled(tools.jackson.core.util.JacksonFeature)" because "this._readCapabilities" is null
	at tools.jackson.databind/tools.jackson.databind.DeserializationContext.isEnabled(DeserializationContext.java:456)
	at tools.jackson.databind/tools.jackson.databind.deser.std.StdDeserializer._checkFromStringCoercion(StdDeserializer.java:1499)
	at tools.jackson.databind/tools.jackson.databind.deser.std.StdDeserializer._checkFromStringCoercion(StdDeserializer.java:1474)
	at tools.jackson.databind/tools.jackson.databind.deser.std.StdDeserializer._parseLong(StdDeserializer.java:985)
	at tools.jackson.databind/tools.jackson.databind.deser.jdk.NumberDeserializers$LongDeserializer.deserialize(NumberDeserializers.java:626)
	at tools.jackson.databind/tools.jackson.databind.deser.jdk.NumberDeserializers$LongDeserializer.deserialize(NumberDeserializers.java:591)
	at tools.jackson.databind/tools.jackson.databind.DeserializationContext._readValue(DeserializationContext.java:406)
	at tools.jackson.databind/tools.jackson.databind.DeserializationContext.readValue(DeserializationContext.java:390)
	at tools.jackson.databind/tools.jackson.databind.DeserializationContext.readValue(DeserializationContext.java:366)
	at tools.jackson.core@3.1.0-SNAPSHOT/tools.jackson.core.base.ParserMinimalBase.readValueAs(ParserMinimalBase.java:810)
	at tools.jackson.databind/tools.jackson.databind.Kotlin1089Test$Inner$Deser.deserialize(Kotlin1089Test.java:20)
	at tools.jackson.databind/tools.jackson.databind.Kotlin1089Test$Inner$Deser.deserialize(Kotlin1089Test.java:15)
	at tools.jackson.databind/tools.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:569)
	at tools.jackson.databind/tools.jackson.databind.deser.bean.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:858)
	... 12 more
```

### Expected behavior

_No response_

### Additional context

This is due to the following issue reported to `kotlin-module`.
https://github.com/FasterXML/jackson-module-kotlin/issues/1089

--- a/release-notes/VERSION
+++ b/release-notes/VERSION
@@ -213,7 +213,7 @@ Versions: 3.x (for earlier see VERSION-2.x)
 #5685: Fix issue where `_findSerializer()` was called with incorrect argument in
   `StdDelegatingSerializer::isEmpty()`
  (fixed by @wrongwrong)
-#5686: ix NPE where code can look for an objectId when no `_objectIdReader` is set
+#5686: Fix NPE where code can look for an objectId when no `_objectIdReader` is set
  (fix by @pjfanning)
 
 3.0.4 (21-Jan-2026)

--- a/release-notes/VERSION
+++ b/release-notes/VERSION
@@ -207,6 +207,9 @@ Versions: 3.x (for earlier see VERSION-2.x)
  (implemented by @pjfanning)
 #5674: Fix NPE when deserializing `Throwable` with `null` "stackTrace"
  (fix by @pjfanning)
+#5683: When deserializing using `ObjectMapper::treeToValue`, NPE occurs
+  when `JsonParser::readValueAs` is called
+ (reported by @wrongwrong)
 #5685: Fix issue where `_findSerializer()` was called with incorrect argument in
   `StdDelegatingSerializer::isEmpty()`
  (fixed by @wrongwrong)

--- a/src/main/java/tools/jackson/databind/ObjectMapper.java
+++ b/src/main/java/tools/jackson/databind/ObjectMapper.java
@@ -888,8 +888,8 @@ public class ObjectMapper
     @Override
     public JsonParser treeAsTokens(JsonNode n) {
         _assertNotNull("n", n);
-        DeserializationContext ctxt = _deserializationContext();
-        return new TreeTraversingParser(n, ctxt);
+        DeserializationContextExt ctxt = _deserializationContext();
+        return ctxt.assignAndReturnParser(new TreeTraversingParser(n, ctxt));
     }
 
     /**

--- a/src/main/java/tools/jackson/databind/ObjectReader.java
+++ b/src/main/java/tools/jackson/databind/ObjectReader.java
@@ -1104,9 +1104,9 @@ public class ObjectReader
         return treeAsTokens(n, _deserializationContext());
     }
 
-    protected JsonParser treeAsTokens(JsonNode n, DeserializationContext ctxt) {
+    protected JsonParser treeAsTokens(JsonNode n, DeserializationContextExt ctxt) {
         _assertNotNull("n", n);
-        return new TreeTraversingParser(n, ctxt);
+        return ctxt.assignAndReturnParser(new TreeTraversingParser(n, ctxt));
     }
 
     /**

--- a/src/test/java/tools/jackson/databind/records/RecordViaParser5683Test.java
+++ b/src/test/java/tools/jackson/databind/records/RecordViaParser5683Test.java
@@ -1,0 +1,53 @@
+package tools.jackson.databind.records;
+
+import org.junit.jupiter.api.Test;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+
+import tools.jackson.core.*;
+
+import tools.jackson.databind.*;
+import tools.jackson.databind.annotation.JsonDeserialize;
+import tools.jackson.databind.deser.std.StdDeserializer;
+import tools.jackson.databind.testutil.DatabindTestUtil;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RecordViaParser5683Test extends DatabindTestUtil
+{
+    @JsonDeserialize(using = Inner.Deser.class)
+    public record Inner(@JsonValue long value) {
+        static class Deser extends StdDeserializer<Inner> {
+            protected Deser() { super(Inner.class); }
+
+            @Override
+            public Inner deserialize(JsonParser p,  DeserializationContext ctxt) {
+                return new Inner(p.readValueAs(Long.class));
+            }
+        }
+    }
+
+    public record Outer(Inner inner) { }
+
+    /*
+    /**********************************************************************
+    /* Test methods
+    /**********************************************************************
+     */
+
+    private final ObjectMapper MAPPER = newJsonMapper();
+    
+    @Test
+    public void testIssue5683()
+    {
+        final String json = "{\"inner\":\"123\"}";
+        Outer value;
+
+        value = MAPPER.readValue(json, Outer.class);
+        assertEquals(123L, value.inner.value());
+
+        JsonNode tree = MAPPER.readTree(json);
+        value = MAPPER.treeToValue(tree, Outer.class);
+        assertEquals(123L, value.inner.value());
+    }
+}

--- a/src/test/java/tools/jackson/databind/records/RecordViaParser5683Test.java
+++ b/src/test/java/tools/jackson/databind/records/RecordViaParser5683Test.java
@@ -41,13 +41,17 @@ public class RecordViaParser5683Test extends DatabindTestUtil
     public void testIssue5683()
     {
         final String json = "{\"inner\":\"123\"}";
+        final JsonNode tree = MAPPER.readTree(json);
+
         Outer value;
 
         value = MAPPER.readValue(json, Outer.class);
         assertEquals(123L, value.inner.value());
 
-        JsonNode tree = MAPPER.readTree(json);
         value = MAPPER.treeToValue(tree, Outer.class);
+        assertEquals(123L, value.inner.value());
+
+        value = MAPPER.reader().treeToValue(tree, Outer.class);
         assertEquals(123L, value.inner.value());
     }
 }

--- a/src/test/java/tools/jackson/databind/ser/filter/JsonIncludeForArray5515Test.java
+++ b/src/test/java/tools/jackson/databind/ser/filter/JsonIncludeForArray5515Test.java
@@ -15,9 +15,9 @@ public class JsonIncludeForArray5515Test
         extends DatabindTestUtil
 {
     /*
-    /**********************************************************
+    /**********************************************************************
     /* Filters
-    /**********************************************************
+    /**********************************************************************
      */
 
     static class Foo5515Filter {
@@ -31,9 +31,9 @@ public class JsonIncludeForArray5515Test
     }
 
     /*
-    /**********************************************************
+    /**********************************************************************
     /* POJOs — one per array type
-    /**********************************************************
+    /**********************************************************************
      */
 
     static class ObjectArray5155Pojo {
@@ -101,20 +101,14 @@ public class JsonIncludeForArray5515Test
         }
     }
 
-    /*
-    /**********************************************************
-    /* Mapper
-    /**********************************************************
-     */
-
     private final ObjectMapper MAPPER = jsonMapperBuilder()
             .enable(SerializationFeature.APPLY_JSON_INCLUDE_FOR_CONTAINERS)
             .build();
 
     /*
-    /**********************************************************
-    /* Tests — reference arrays (expect filtering, FAIL today)
-    /**********************************************************
+    /**********************************************************************
+    /* Tests — reference arrays (expect filtering)
+    /**********************************************************************
      */
 
     @Test
@@ -122,8 +116,6 @@ public class JsonIncludeForArray5515Test
         ObjectArray5155Pojo input = new ObjectArray5155Pojo(
                 "1", "foo", "2"
         );
-
-        // EXPECT foo to be filtered out — FAILS today
         assertEquals(
                 a2q("{'values':['1','2']}"),
                 MAPPER.writeValueAsString(input)
@@ -155,9 +147,9 @@ public class JsonIncludeForArray5515Test
     }
 
     /*
-    /**********************************************************
-    /* Tests — primitive arrays (expect NON_DEFAULT filtering, FAIL today)
-    /**********************************************************
+    /**********************************************************************
+    /* Tests — primitive arrays (expect NON_DEFAULT filtering)
+    /**********************************************************************
      */
 
     @Test
@@ -166,7 +158,6 @@ public class JsonIncludeForArray5515Test
                 true, false, true
         );
 
-        // EXPECT default 'false' to be filtered out — FAILS today
         assertEquals(
                 a2q("{'values':[true,true]}"),
                 MAPPER.writeValueAsString(input)
@@ -179,7 +170,6 @@ public class JsonIncludeForArray5515Test
                 0, 1, 0, 2
         );
 
-        // EXPECT default '0' to be filtered out — FAILS today
         assertEquals(
                 a2q("{'values':[1,2]}"),
                 MAPPER.writeValueAsString(input)
@@ -191,8 +181,6 @@ public class JsonIncludeForArray5515Test
         LongArray5515Pojo input = new LongArray5515Pojo(
                 0L, 1L, 0L, 2L
         );
-
-        // EXPECT default '0L' to be filtered out — FAILS today
         assertEquals(
                 a2q("{'values':[1,2]}"),
                 MAPPER.writeValueAsString(input)
@@ -204,8 +192,6 @@ public class JsonIncludeForArray5515Test
         DoubleArray5515Pojo input = new DoubleArray5515Pojo(
                 0.0, 1.5, 0.0
         );
-
-        // EXPECT default '0.0' to be filtered out — FAILS today
         assertEquals(
                 a2q("{'values':[1.5]}"),
                 MAPPER.writeValueAsString(input)

